File: questPresentation/src/main/java/ru/nekit/android/qls/setupWizard/view/UnlockSecretFragment.java
package ru.nekit.android.qls.setupWizard.view;

import android.support.annotation.NonNull;
import android.view.View;

import com.andrognito.patternlockview.PatternLockView;
import com.andrognito.patternlockview.listener.PatternLockViewListener;
import com.andrognito.patternlockview.utils.PatternLockUtils;

import java.util.List;

import ru.nekit.android.qls.R;
import ru.nekit.android.qls.setupWizard.BaseSetupWizard;
import ru.nekit.android.qls.setupWizard.BaseSetupWizardStep;
import ru.nekit.android.qls.setupWizard.ISetupWizardStep;
import ru.nekit.android.qls.utils.KeyboardHost;
import ru.nekit.android.qls.utils.Vibrate;

import static com.andrognito.patternlockview.PatternLockView.PatternViewMode.WRONG;
import static ru.nekit.android.qls.setupWizard.QuestSetupWizard.QuestSetupWizardStep.SETUP_UNLOCK_SECRET;

public class UnlockSecretFragment extends QuestSetupWizardFragment
        implements PatternLockViewListener {

    private ISetupWizardStep mStep;
    private PatternLockView mPatternLockView;
    private List<PatternLockView.Dot> mPattern;

    public static UnlockSecretFragment getInstance(@NonNull ISetupWizardStep step) {
        UnlockSecretFragment fragment = new UnlockSecretFragment();
        fragment.mStep = step;
        return fragment;
    }

    @Override
    public void onStarted() {

    }

    @Override
    public void onProgress(List<PatternLockView.Dot> pattern) {
        if (mStep == SETUP_UNLOCK_SECRET) {
            setAltButtonVisibility(pattern.size() >= BaseSetupWizard.UNLOCK_SECRET_MIN_SIZE);
        }
    }

    @Override
    public void onResume() {
        KeyboardHost.hideKeyboard(getActivity().getApplicationContext(), mPatternLockView);
        super.onResume();
    }

    @Override
    public void onComplete(List<PatternLockView.Dot> pattern) {
        if (pattern.size() < BaseSetupWizard.UNLOCK_SECRET_MIN_SIZE) {
            mPatternLockView.setViewMode(WRONG);
        } else {
            if (mStep == SETUP_UNLOCK_SECRET) {
                mPattern = pattern;
                update(true);
            } else if (mStep == BaseSetupWizardStep.UNLOCK_SECRET) {
                if (getSetupWizard().checkUnlockSecret(
                        PatternLockUtils.patternToMD5(mPatternLockView, pattern))) {
                    showNextSetupWizardStep();
                } else {
                    mPatternLockView.setViewMode(WRONG);
                    Vibrate.make(getContext(), 400);
                }
            }
        }
    }

    @Override
    public void onCleared() {
        resetPattern();
    }

    @Override
    protected void onSetupStart(@NonNull View view) {
        mPatternLockView = (PatternLockView) view.findViewById(R.id.unlock_secret_view);
        mPatternLockView.addPatternLockListener(this);
        mPatternLockView.setDotCount(3);
        mPatternLockView.setAspectRatioEnabled(true);
        mPatternLockView.setDotAnimationDuration(150);
        mPatternLockView.setPathEndAnimationDuration(100);
        mPatternLockView.setInStealthMode(false);
        mPatternLockView.setTactileFeedbackEnabled(true);
        mPatternLockView.setInputEnabled(true);
        if (mStep == SETUP_UNLOCK_SECRET) {
            setNextButtonText(R.string.label_set_unlock_secret);
            setAltButtonText(R.string.label_reset_unlock_secret);
        } else if (mStep == BaseSetupWizardStep.UNLOCK_SECRET) {
            setToolContainerVisibility(false);
        }
        setAltButtonVisibility(false);
        update(false);
    }

    public void onDestroy() {
        mPatternLockView.removePatternLockListener(this);
        super.onDestroy();
    }

    @Override
    protected boolean nextAction() {
        getSetupWizard().setUnlockSecret(PatternLockUtils.patternToMD5(mPatternLockView, mPattern));
        return true;
    }

    @Override
    protected int getLayoutId() {
        return R.layout.sw_unlock_secret;
    }

    private void update(boolean unlockSecretIsSet) {
        setNextButtonVisibility(unlockSecretIsSet);
    }

    @Override
    protected void altAction() {
        mPatternLockView.clearPattern();
        resetPattern();
    }

    private void resetPattern() {
        mPattern = null;
        setAltButtonVisibility(false);
        setNextButtonVisibility(false);
    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/setupWizard/view/UnlockSecretFragment.kt
package ru.nekit.android.qls.setupWizard.view

import android.os.Bundle
import android.support.annotation.LayoutRes
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.RelativeLayout
import com.andrognito.patternlockview.PatternLockView
import com.andrognito.patternlockview.PatternLockView.PatternViewMode.WRONG
import com.andrognito.patternlockview.listener.PatternLockViewListener
import com.andrognito.patternlockview.utils.PatternLockUtils
import io.reactivex.Single
import ru.nekit.android.qls.R
import ru.nekit.android.qls.setupWizard.BaseSetupWizard
import ru.nekit.android.qls.setupWizard.BaseSetupWizardStep
import ru.nekit.android.qls.setupWizard.BaseSetupWizardStep.UNLOCK_SECRET
import ru.nekit.android.qls.setupWizard.ISetupWizardStep
import ru.nekit.android.qls.setupWizard.QuestSetupWizard.QuestSetupWizardStep.SET_UNLOCK_SECRET
import ru.nekit.android.utils.Delay
import ru.nekit.android.utils.KeyboardHost
import ru.nekit.android.utils.ParameterlessSingletonHolder
import ru.nekit.android.utils.Vibrate

class UnlockSecretFragment : QuestSetupWizardFragment(), PatternLockViewListener {

    var step: ISetupWizardStep? = null
    private lateinit var patternLockView: PatternLockView
    private lateinit var pattern: List<PatternLockView.Dot>

    override fun onStarted() {
    }

    override fun onProgress(pattern: List<PatternLockView.Dot>) {
        if (step == SET_UNLOCK_SECRET) {
            update(pattern.size >= BaseSetupWizard.UNLOCK_SECRET_MIN_SIZE, false)
        }
    }

    override fun onResume() {
        KeyboardHost.hideKeyboard(context!!, patternLockView, Delay.KEYBOARD.get(context!!))
        super.onResume()
    }

    override fun onComplete(pattern: List<PatternLockView.Dot>) {
        if (pattern.size < BaseSetupWizard.UNLOCK_SECRET_MIN_SIZE) {
            patternLockView.setViewMode(WRONG)
        } else {
            if (step == SET_UNLOCK_SECRET) {
                this.pattern = pattern
                update(true, true)
            } else if (step == UNLOCK_SECRET) {
                setupWizard.checkUnlockSecret(PatternLockUtils.patternToMD5(patternLockView, pattern)) { it ->
                    if (it) {
                        patternLockView.clearPattern()
                        goNext()
                    } else {
                        patternLockView.setViewMode(WRONG)
                        Vibrate.make(context!!, 400)
                    }
                }
            }
        }
    }

    override fun onCleared() {
        resetPattern()
    }

    override fun onSetupStart(view: View) {
        patternLockView = view.findViewById<View>(R.id.unlock_view) as PatternLockView
        with(patternLockView) {
            addPatternLockListener(this@UnlockSecretFragment)
            dotCount = 3
            isAspectRatioEnabled = true
            dotAnimationDuration = 150
            pathEndAnimationDuration = 100
            isInStealthMode = false
            isTactileFeedbackEnabled = true
            isInputEnabled = true
        }
        if (step == SET_UNLOCK_SECRET) {
            title = R.string.title_setup_unlock_secret
            nextButtonText(R.string.label_set_unlock_secret)
            altButtonText(R.string.label_reset_unlock_secret)
            descriptionView.visibility = VISIBLE
        } else if (step == UNLOCK_SECRET) {
            title = R.string.title_enter_unlock_secret
            toolContainerVisibility(false)
            descriptionView.visibility = GONE
            (patternLockView.layoutParams as RelativeLayout.LayoutParams).apply {
                addRule(RelativeLayout.CENTER_IN_PARENT, RelativeLayout.TRUE)
                patternLockView.layoutParams = this
            }
        }
        update(false, false)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        step?.apply {
            outState.putString(CURRENT_STEP, name)
        }
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        savedInstanceState?.let {
            step = BaseSetupWizardStep.valueOf(savedInstanceState.getString(CURRENT_STEP))
        }
        super.onViewStateRestored(savedInstanceState)
    }

    override fun onDestroy() {
        patternLockView.removePatternLockListener(this)
        super.onDestroy()
    }

    override fun nextAction(): Single<Boolean> =
            setupWizard.setUnlockSecret(PatternLockUtils.patternToMD5(patternLockView, pattern)).toSingleDefault(true)

    @LayoutRes
    override fun getLayoutId() = R.layout.sw_unlock_secret

    private fun update(unlockSecretCanBeReset: Boolean, unlockSecretIsSet: Boolean) {
        toolContainerVisibility(unlockSecretCanBeReset || unlockSecretIsSet)
        altButtonVisibility(unlockSecretCanBeReset)
        nextButtonVisibility(unlockSecretIsSet)
    }

    override fun altAction() {
        patternLockView.clearPattern()
        resetPattern()
    }

    private fun resetPattern() {
        pattern = ArrayList()
        update(false, false)
    }

    companion object : ParameterlessSingletonHolder<UnlockSecretFragment>(::UnlockSecretFragment) {

        const val CURRENT_STEP = "current_step"

    }
}